### PR TITLE
add hack/ helpers for bootstrapping engine image using stable SDK

### LINKED ISSUE #3739: Update `engine:dev` to be more of a bootstrap
From conversation in discord:
1. We can use stable dagger to build dev dagger, but we should be sure that this "bootstrap" step only uses a stable sdk + engine. It thus might need to have its own `go.mod` (maybe can use the [`tools.go` convention](https://play-with-go.dev/tools-as-dependencies_go119_en/) here?) so we can ensure its only using stable versions.
2. Our CI steps will typically do something like:
   1. Build+start dev dagger (bootstrapping it with stable dagger)
   1. Export DAGGER_HOST=docker-container://test-dagger-engine
   1. Run the rest of our CI steps
3. But developers locally can retain the flexibility of doing the above or instead using stable dagger to run ci steps (i.e. I broke my local dev engine but want to lint my python code still)
4. The above means that the bootstrap stage where we build dagger w/ dagger can only rely on stable released features. The rest of the steps can use dev features though. I think that's probably fine? Worst case scenario it's always possible to add more stages to your bootstrap process...

Additionally, it would make sense to find a way of modelling this as a sort of "wrapper" where you can run a command that bootstraps dev dagger and then invokes the rest of the commandline you provide in the context of that dev dagger (i.e. `DAGGER_HOST=...`.

A kind of silly approach to this could be `mage devdagger sdk:python:test`, where the `devdagger` target invokes mage as a subprocess. But there's probably better ways of doing it too.

cc @aluzzardi @shykes @vito

### PR DESCRIPTION
**Pull request by vito**
fixes #3739 

* track bootstrap SDK version in separate Go module
* move mage code under magefiles/ so it can be optionally run with its own go.mod/go.sum without clashing with the root go.mod
  * while also automatically picked up by regular `mage` CLI when you don't care about bootstrapping
* run `mage boot:run TARGET` to run a target using bootstrap SDK version
  * e.g. `mage boot:run engine:dev`
* run `mage boot:dev TARGET` to run a target using dev engine image built using bootstrap SDK version
  * e.g. `mage boot:dev engine:test`

### CODE REVIEW

**Review by sipsma — COMMENTED** [review]
Had some more ideas about approaches here, just kind of spitballing.

Given the setup from this PR: https://github.com/dagger/dagger/pull/3756

What do you think about a slightly different approach like this:
1. `./hack/dev` is a script that builds dev dagger using the bootstrap sdk
   * If you provide no args, all it does is build dev dagger, run it in docker and print out the `DAGGER_HOST` to use to connect to it
   * If you do provide args it executes whatever they are with the `DAGGER_HOST` env set
     * `./hack/dev make ...` would let you run any other automation with your dev engine (this might also require that ./hack/dev adds the `hack` dir to $PATH before exec'ing the args)
     * `./hack/dev bash` would open a shell where all your commands run against the dev engine
2. Rather than symlinking go.mod, `./hack/dev` just does a `cd` into the directory where the other go.mod exists and does `go run` from there to build and deploy the dev engine.
   * I just suspect this might be simpler and possible in this different approach, but could be missing something

I think that might be more flexible than the original idea of fork/exec'ing mage.

The biggest downside I can think of is that it still means users would have to remember to do `./hack/dev ...` in order to run against the dev engine rather than it being the default (and running against not dev being the exception), which I think was one of @aluzzardi's general concerns around all this stuff too.
* `./hack/dev bash` helps with this I suppose. Maybe we could view `./hack/dev` as the general entrypoint to doing any sort of local dev? Or maybe that's really limiting, not sure.

Just curious to get your initial thoughts on this, not so strongly opinionated that I think it's worth blocking this for a long time bikeshedding the DX of our `./hack` dir :-)

**Comment by vito** [comment]
@sipsma Makes sense, yeah this feels like a good time for just a bit of Bash glue instead. I'll play around with the new setup.

I'll likely introduce a `hack/test` since I find this to be a good sane default entrypoint; it'll probably just be shorthand for `hack/dev hack/make engine:test` or something.

**Comment by aluzzardi** [comment]
Something is bothering me, but I can't quite put my finger on it. Maybe that's just the price to pay to test dagger with dagger, I can't find a simpler approach.

My main concern is our tests go through 3 different DAGGER_HOST -- 1) bootstrap with release engine 2) run pipeline with bootstrapped engine 3) dagger-in-dagger for tests running in a ctr.Exec()

Even though I know how this works, I keep getting confused ("where am I?")

Trying to find an alternative but I'm empty handed for now

**Comment by sipsma** [comment]
> My main concern is our tests go through 3 different DAGGER_HOST -- 1) bootstrap with release engine 2) run pipeline with bootstrapped engine 3) dagger-in-dagger for tests running in a ctr.Exec()

@aluzzardi I think this is just what happens when you dogfood and bootstrap though. In my experience, the bootstrap builds for compilers and distributions is always confusing+mindblowing. Honestly even more so than what we're doing since it'll often involve even more stages. So while yeah I agree that building dagger w/ dagger and testing dagger in dagger gets meta+confusing at times, it's not necessarily indicative that we're doing anything wrong. I think it's par for the course.

The alternative is what @vito said, use something else to build and test dagger. We can of course, I just personally think we'd be resigning to that too soon and missing out on great dogfooding opportunities.
* E.g. the engine image is going to grow in complexity more and more as time goes on I think. So we'll need to rely on a progressively more and more complicated Dockerfile or will have to break up the bootstrap into multiple stages where you first build a simple engine w/ Dockerfile and use that to build the more complicated one, which is just as complex if not worse than just building with dagger in the first place.

---

In terms of the build, I really think we can isolate the complication entirely to just the first part where you build a dev dagger. The `./hack/dev` (or `./hack/test`) approach seems like it accomplishes that. It can also just consist of a single isolated directory where there's a go.mod and you just run the build from there, done. After that you have your dev dagger running and usable by future builds+tests.

**Comment by vito** [comment]
@sipsma @aluzzardi I'm good to carry on with the dogfooding. I went through similar confusion with Bass (see [`buildkit.bass`](https://github.com/vito/bass/blob/99973243f585be969e32909669405aaad1948f2c/bass/buildkit.bass) for my own image building code) and it definitely became handy to have a programmable builder for the Buildkit image.

The part where it was most helpful was being able to configure IP address ranges for CNI so the inner test Buildkit instance didn't conflict with the outer one ([here](https://github.com/vito/bass/blob/main/bass/buildkit.bass#L71-L83)).

My case was a little different since I'm actually running another Buildkit as a service with Bass and then running tests against it, whereas our Dagger-in-Dagger just forwards the outer engine through. But the principle still stands that our engine image will likely beef up someday and it'll be nice to have Dagger at that point.

---

Latest changes should be ready to review. I ended up having to keep the restructure (`internal/mage/*` => `./magefiles`) and `go.mod` symlink dance for the same reason as before: once Go sees a `go.mod` exists it treats it as a module boundary regardless of where you run it from. So the root `magefile.go` has to move elsewhere, and luckily Mage also respects `magefiles/`, which is a convenient place to install the temporary `go.mod`.

Happy to take another swing at it though if I missed something.

File: .github/workflows/lint.yml
Comment (sipsma) [inline .github/workflows/lint.yml:21]: Revisiting this now, one last tweak I think might make sense is something like this:
* `./hack/make engine:test` - results in the dev engine being bootstrapped and running the provided target
   * basically, `./hack/make` always implicitly wraps itself with `./hack/dev` (or the end effect of doing that)
* `./hack/make-prod engine:test` results in the tests running against the actual prod release, basically just skip the dev bootstrap

We can still retain `./hack/dev` too to support wrapping arbitrary commands.

Comment (aluzzardi) [reply .github/workflows/lint.yml:21]: Ran into a bunch of problems recently (catch 22): https://github.com/dagger/dagger/pull/3928#discussion_r1027021899

Brainstorming here, but would this be simpler:

- We write our pipelines targeting stable *entirely*
  - separate `go.mod`, there's no multiple version of dagger in the same codebase competing against each other
- In our pipelines, we bootstrap dagger and run tests with dagger-in-dagger-main
  - `internal/mage` -> stable
  - `core/integration`, `sdk/*/*test` -> main

Comment (vito) [reply .github/workflows/lint.yml:21]: While we're talking about this, I'd also _really_ like the ability to run the tests against a native local engine. Having to rebuild the image all the time for engine changes (even just `log.Println`) is a pretty gnarly hit to the feedback loop. :confused:

I'm still for testing against the engine image since that's what actually ships now, and maybe we do that in CI where the feedback loop doesn't matter, but at least for local dev it'd be awesome to optimize.

**Comment by sipsma** [comment]
@vito @aluzzardi I got hard blocked by a need for bootstrapping in this PR: https://github.com/dagger/dagger/pull/3983

So I went ahead and rebased this and applied the simplifications discussed in comments:
1. *All* mage steps use the separate go.mod now, no script for symlinking them around or anything
2. `./hack/make-prod` runs the mage steps using the stable dagger specified in that go.mod
3. `./hack/make` bootstraps `docker-container://test-dagger-engine` from local code and then runs mage steps
   * It's thus the same as before, but no need for our mage code to wrap everything in `WithDaggerEngine`, that util is gone entirely
3. `./hack/dev` bootstraps `docker-container://test-dagger-engine` and then runs any command you want with the `DAGGER_HOST` and other env vars set to point to it
   * @vito I *think* that `./hack/dev bash` (or replace bash with whatever shell) enables the "run the tests against a native local engine" use case you mentioned above. Basically, run that command and now in your shell you'll be able to just run tests natively without rebuilding the engine every time. Let me know if that's what you had in mind or if it was something else.

There's also docs for all this in `hack/README.md` now.

Open to bikeshedding on the various name choices, I have no attachment to the choices I went with.

**Comment by sipsma** [comment]
Another nice side effect of the updates - CI is probably slightly faster because we use `./hack/make-prod` for linting, which lets us skip rebuilding the engine just to run some linting. cc @gerhard

**Comment by sipsma** [comment]
Okay after writing my last comment I went back to use this on #3983 and realized there was *still* a catch 22.

Fortunately I found a fix that's relatively painless, pushed as a separate commit. Explanation below.

---

The problem is that PR changes engine session to communicate over unix sockets, which means that engine-session binary (and thus the engine image) changes **AND** the Go SDK needs to change in order communicate with it.

With the change I made here earlier, I could successfully bootstrap the dev engine despite those backwards incompatible changes because we now used stable engine to build dev engine. ***But*** then when I went to actually run the tests it failed because it was trying to use the *old* go sdk with the new dev engine-session.

Basically, this situation called for something like the original approach of symlinking in go.mod opportunistically so that we could switch between running the CI first with stable go sdk and then later with dev go sdk. However, I found a pretty nice hack to avoid that: `go.work`. Go workspaces have two extremely convenient properties:
1. They enable `replace` statements that work the same as go.mod replaces
2. They can be easily disabled by just setting `GOWORK=off`

The above gives us the ability to switch back and forth between use of the stable go sdk and the local dev go sdk by just switching `GOWORK` on and off.

***TO BE VERY CLEAR: all of the above nonsense is taken care of by the hack scripts***. You and anyone else shouldn't have to think about it at all day-to-day. That being said, I documented all this in `./hack/README.md` for anyone hacking on ./hack in the future.

---

@vito @aluzzardi I don't like the complication of the setup still but provided there isn't any functional issue with this approach I think we should merge it as an incremental improvement:
1. It solves the bootstrapping problem; #3983 is now buildable+testable whereas before it wasn't
2. I don't think it's actually more complicated to think about than what we had going on previously, it's pretty much just as bad in that regard

**Comment by marcosnils** [comment]
> There's also docs for all this in `hack/README.md` now. cc @marcosnils since you mentioned writing docs for building the CLI locally recently, it may fit in that file.

:+1:  I'll write something and send a PR targeting this branch as base.

**Comment by vito** [comment]
Nice, I like the use of Go workspaces for this, the `GOWORK=off` toggle makes it much less awkward than the symlink dance.

> 4. `./hack/dev` bootstraps `docker-container://test-dagger-engine` and then runs any command you want with the `DAGGER_HOST` and other env vars set to point to it
> 
>     * @vito I _think_ that `./hack/dev bash` (or replace bash with whatever shell) enables the "run the tests against a native local engine" use case you mentioned above. Basically, run that command and now in your shell you'll be able to just run tests natively without rebuilding the engine every time. Let me know if that's what you had in mind or if it was something else.

Not quite, I still have to run `./hack/dev` to start the new engine image if I make changes to e.g. `core/*.go`, unless I'm missing something. This definitely isn't a blocker though. Happy to merge your changes since this ship has already sailed, though it looks like it needs rebase.

**Comment by sipsma** [comment]
@vito thanks for the rebase, was just about to do that and saw it was already done :-)

@aluzzardi let me know if you have any concerns with this approach

**Comment by aluzzardi** [comment]
I think I suggested this in another PR but can't find the link: Wouldn't it be simpler and solve all the issues if we just always write our internal pipelines using stable?

As in, we put a `go.mod` in `internal/mage` and the pipelines are **always** targeting *a* stable release. Whereas our tests **always** target *main* (root `go.mod` has a `replace`)

> * `./hack/make-prod` runs the mage steps using the stable dagger
> * `./hack/make` bootstraps `docker-container://test-dagger-engine` from local code and then runs mage steps

I don't fully understand when we'd use one rather than the other. For instance, I got recently stuck with `Stdout` returning `string` rather than a `File`. In that case, either the pipelines assume a `string` is returned then only `./hack/make` works, or they assume a `File` is returned and only `./hack/make-prod` works, but they're not interchangeable right?

**Comment by sipsma** [comment]
> As in, we put a go.mod in internal/mage and the pipelines are always targeting a stable release. Whereas our tests always target main (root go.mod has a replace)

That's what I did at first too in [this commit](https://github.com/dagger/dagger/pull/3755/commits/12a0b3c2f7b2405cc3999bc26d1d213b131746a6) but realized it doesn't work in every case.

Using the example of my other PR where I am updating SDKs to communicate with engine session over unix sockets, this is what happens if you just have a separate `internal/mage/go.mod` where a stable go sdk is specified and then try to run `engine:test`:
1. You successfully use stable Go SDK + stable engine to bootstrap dev engine
2. You then try to run the engine tests against it, which requires dagger-in-dagger and thus setting up an exec
3. You are now trying to use the stable Go SDK to run an exec with the dev engine. Stable Go SDK tries speaking over localhost to engine-session using unix sockets and you get an error.

The go.work approach fixes this by allowing us to use stable sdk + stable engine to bootstrap and then switch to dev sdk + dev engine after the bootstrap.

There are other possible fixes but they are much more involved:
1. Try to update my PR to support both localhost and unix socket connections and make it configurable; but that would be a pretty big change and require weird hacks in our actual code solely for the purpose of working around this bootstrapping issue.
   * Also will need to repeat this approach for any similar changes in the future
2. Support mounting docker.sock into the test container and then mount in the dev engine-session binary instead of using the `PrivilegedNesting` flag.

> For instance, I got recently stuck with Stdout returning string rather than a File. In that case, either the pipelines assume a string is returned then only ./hack/make works, or they assume a File is returned and only ./hack/make-prod works, but they're not interchangeable right?

Yeah true, this PR fixes the problem where we have an internal backwards incompatible change (unix sockets in my case) but it won't work if we ever need to make another fully backwards-incompatible API change like that (though `@deprecated` APIs should still be fine).

@aluzzardi What do you think of this plan:
1. Use go.work for now so the other PR is unblocked
   * I don't think we have any fully backwards incompatible changes in the API planned, so this could be okay on a very temporary basis?
2. Once we have docker socket mount support then switch our engine tests to use that
3. At that time, also change back to my previous commit where there is a separate `go.mod` for `internal/mage`, no more switching back and forth with `go.work`.

If there's other approaches for unblocking my PR I'm all ears. I just really want to avoid that PR needing to introduce a way of supporting both localhost and unix sockets at the same time, but if others think making that change is better than the go.work approach (and associated api backcompat issues) then I can go that route instead.

**Comment by aluzzardi** [comment]
> @aluzzardi What do you think of this plan:
> 
> 1. Use go.work for now so the other PR is unblocked
>    
>    * I don't think we have any fully backwards incompatible changes in the API planned, so this could be okay on a very temporary basis?
> 2. Once we have docker socket mount support then switch our engine tests to use that
> 3. At that time, also change back to my previous commit where there is a separate `go.mod` for `internal/mage`, no more switching back and forth with `go.work`.

Yes, that makes sense! Whatever unblocks us right away.

Long term, I think a separate `go.mod` is the "safe" plan since there's still chicken-and-egg problems (e.g. `stdout`)

> Yeah true, this PR fixes the problem where we have an internal backwards incompatible change (unix sockets in my case) but it won't work if we ever need to make another fully backwards-incompatible API change like that (though @deprecated APIs should still be fine).

One cool but annoying side effect of `@deprecated` is the linter fails when using `@deprecated` functions. I don't know whether linters honor `go.work` or not, but either way a `@deprecated` usage is still a "breaking" change for the CI

> * You then try to run the engine tests against it, which requires dagger-in-dagger and thus setting up an exec
> * You are now trying to use the stable Go SDK to run an exec with the dev engine. Stable Go SDK tries speaking over localhost to engine-session using unix sockets and you get an error.

Not sure I fully understand this part. I would have assumed that mage uses stable (sdk and engine) all the way up until `exec(go test)` (and mounting the dev session binary), at which point we'd "switch" to dev for the duration of the tests. The problem though is the session binary is talking to non-dev engine and that's where it falls apart. Is that correct?

**Comment by sipsma** [comment]
> I would have assumed that mage uses stable (sdk and engine) all the way up until exec(go test) (and mounting the dev session binary), at which point we'd "switch" to dev for the duration of the tests. 

That's what would be ideal, but in practice we can't do that yet. In order to test against the dev engine we have to set `DAGGER_HOST=docker-container://test-dagger-engine`, but if we do that then the stable SDK is trying to setup the dagger-in-dagger exec using the dev session binary and it fails. So in this case it's that we have a backwards incompatible change between SDK <-> session bin.

That's why support for mounting `docker.sock` should fix it; then we don't have to set `DAGGER_HOST` on the host; we only have to set it in the exec and can just use stable sdk on the host the whole time.

**Comment by aluzzardi** [comment]
> That's what would be ideal, but in practice we can't do that yet. In order to test against the dev engine we have to set `DAGGER_HOST=docker-container://test-dagger-engine`, but if we do that then the stable SDK is trying to setup the dagger-in-dagger exec using the dev session binary and it fails. So in this case it's that we have a backwards incompatible change between SDK <-> session bin.

In the "new world" approach (https://github.com/dagger/dagger/issues/3830#issuecomment-1325661323), we could just set `_EXPERIMENTAL_DAGGER_CLI_BIN` and mount the binary, right? The only problem though is it would be talking to "stable" engine, assuming the shim sets up `_EXPERIMENTAL_DAGGER_RUNNER_HOST`

**Comment by sipsma** [comment]
> The only problem though is it would be talking to "stable" engine, assuming the shim sets up _EXPERIMENTAL_DAGGER_RUNNER_HOST

Yeah exactly, we wouldn't actually be testing the engine we want to test.

Once all the changes in #3830 are through and docker.sock mounting is available, engine tests can be updated to do this:
1. Use stable sdk+engine to bootstrap `test-dagger-engine` container (same as today)
2. Instead of setting `ExperimentalPrivilegedNesting` to true in the engine test exec, mount `docker.sock` in there along with the *dev* `engine-session` bin
3. Explicitly set `_EXPERIMENTAL_DAGGER_CLI_BIN=/path/to/dev/engine-session`
4. Explicitly set `_EXPERIMENTAL_DAGGER_RUNNER_HOST=docker-container://test-dagger-engine`

**Comment by aluzzardi** [comment]
Makes sense! :shipit:

**Comment by sipsma** [comment]
(I just saw that the merge commit from before seems to have undone some new logic in the nodejs publish step, fixing, then will merge once I have an approval)

**Review by aluzzardi — APPROVED** [review]

**Comment by sipsma** [comment]
I used the squash and merge button based on recent advice, for some reason that resulted in `undefined` being the commit message? https://github.com/dagger/dagger/commit/58c782e6d5de4992046f2b1ebeea52948ae793fd

Sorry that looks weird @vito, I didn't fully realize the implications of squash+merge.

**Comment by vito** [comment]
lol, weird. np!

there must be some JS in there somewhere. :smile:

**Pull request merged**
